### ISSUE TITLE
Edit Product: Provide option to check if SKU exists for a site

### ISSUE THREAD

**Issue by anitaa1990** [woocommerce/woocommerce-ios#1710]
When updating a product, an error can occur if user enters a SKU for a product that already exists. There is an API endpoint available `/wc/v3/products/&_method=get&sku={sku}` that can be used to verify if a SKU exists for a site.

**Comment by pmusolino** [comment]
@anitaa1990 do you store anything locally on Android from the API request?

**Comment by anitaa1990** [comment]
> @anitaa1990 do you store anything locally on Android from the API request?

@pmusolino, apologies for the late response here. No, we don't store anything locally in this scenario. If the API response is empty, then the assumption is that the SKU is available so we just return a boolean flag in this case. 

Can I ask how iOS is handling error scenario in this case? For instance, when verifying if SKU is available, if there is no internet connection or if the API response results in an error, for some reason, are we allowing the user to update the SKU in this case? Or do we display an error message stating we are not able to verify if SKU exists? I'm not sure what the best user experience here would be. 

If a SKU already exists but the API response returns an error, for some reason, should we allow the user to just update the SKU and then if an error occurs in that case, how would we handle it? 

Sorry  for the ton of questions here but since it's already implemented in iOS just wanted to gain some context 😄

**Comment by jaclync** [comment]
Hey @anitaa1990 never too many questions and great points on the cases you mentioned! @pmusolino added the API validation for SKU in https://github.com/woocommerce/woocommerce-ios/pull/1745 - when the SKU field changes (with throttling), we make an API request to check the SKU and disable the "Done" button (or back navigation and then "Save changes") if the request failed or the SKU is used by another Product.

> if there is no internet connection or if the API response results in an error, for some reason, are we allowing the user to update the SKU in this case? Or do we display an error message stating we are not able to verify if SKU exists?

Based on the code, we consider the SKU update as valid if there is an API error so that the user can still save the inventory settings. I personally think this is better than blocking the inventory settings update, and if the SKU is actually invalid we will still show an error on the Edit Product main screen after saving a Product remotely. What do you think? @anitaa1990 @pmusolino 

I also noticed that on WCiOS we enable the inventory settings to be saved ("Done" button is enabled) before the API request comes back, do you think it'd be better to disable saving the inventory settings whenever we start the API request? Or we can just leave the save action enabled?

**Comment by pmusolino** [comment]
> Based on the code, we consider the SKU update as valid if there is an API error so that the user can still save the inventory settings. I personally think this is better than blocking the inventory settings update, and if the SKU is actually invalid we will still show an error on the Edit Product main screen after saving a Product remotely. What do you think? @anitaa1990 @pmusolino

Exactly, I have the same opinion, I don't think that blocking the inventory settings update is the solution here.

> I also noticed that on WCiOS we enable the inventory settings to be saved ("Done" button is enabled) before the API request comes back, do you think it'd be better to disable saving the inventory settings whenever we start the API request? Or we can just leave the save action enabled?

Maybe yes, it would be appropriate to lock the button, letting the SKU be validated if the API request validates the field or if the API return an error (look at the point above). If we all agree, I will open an issue 👍

**Comment by anitaa1990** [comment]
> Based on the code, we consider the SKU update as valid if there is an API error so that the user can still save the inventory settings. I personally think this is better than blocking the inventory settings update, and if the SKU is actually invalid we will still show an error on the Edit Product main screen after saving a Product remotely. What do you think?

Thanks @jaclync  and @pmusolino for the clarification! I had implemented the same thing on the FluxC side and think a non blocking UI here is the best approach. So if we have an error response when validating SKU and we let the user update anyway and if the SKU is actually invalid, currently we will just display an error SnackBar stating that SKU is invalid. Is this the correct approach? cc @kyleaparker @Garance91540 on your thoughts here.

> I also noticed that on WCiOS we enable the inventory settings to be saved ("Done" button is enabled) before the API request comes back, do you think it'd be better to disable saving the inventory settings whenever we start the API request? Or we can just leave the save action enabled?

I think @kyleaparker mentioned that it would be nice to grey out the `DONE` button if there is a sku error instead of hiding it completely. Does that sound good?
